race-buffer/src/async_reader.rs

```diff
@@ -0,0 +1,286 @@
+/// should have sequential consistency with the memory access of the RaceBuffer writer
+pub trait Snapper<E>
+where
+    E: Entry,
+{
+    /// Error during snapping
+    type Error: std::error::Error;
+
+    /// Take a snapshot of the high 32 bits of the write sequence number of the RaceBuffer
+    fn snap_write_seqn_high(&self) -> Result<u32, Self::Error>;
+    /// Take a snapshot of the low 32 bits of the write sequence number of the RaceBuffer
+    fn snap_write_seqn_low(&self) -> Result<u32, Self::Error>;
+
+    /// Take a snapshot of the high 32 bits of the overwrite sequence number of the RaceBuffer
+    fn snap_overwrite_seqn_high(&self) -> Result<u32, Self::Error>;
+    /// Take a snapshot of the low 32 bits of the overwrite sequence number of the RaceBuffer
+    fn snap_overwrite_seqn_low(&self) -> Result<u32, Self::Error>;
+
+    /// Take a snapshot of the RaceBuffer's backing storage at the given index
+    fn snap_storage(&self, index: usize) -> Result<E, Self::Error>;
+}
+
+/// Used to read from a RaceBuffer asynchronously
+pub struct RaceReader<E, S>
+where
+    E: Entry,
+    S: Snapper<E>,
+{
+    /// Struct for reading writer state
+    snapper: S,
+    /// Sequence number of next entry to be read
+    read_seqn: SeqNum,
+    /// Capacity of backing storage
+    storage_cap: usize,
+    /// Cached prefix, not put into buffer until suffix is successfully read
+    stored_prefix: Option<E>,
+}
+
+impl<E, S> RaceReader<E, S>
+where
+    E: Copy + PartialEq + Entry,
+    S: Snapper<E>,
+{
+    /// Create new RaceReader
+    pub fn new(snapper: S, storage_cap: usize) -> RaceReader<E, S> {
+        RaceReader {
+            snapper,
+            read_seqn: 0.into(),
+            storage_cap,
+            stored_prefix: None,
+        }
+    }
+
+    /// Attempt to read all new entries in race buffer into given read buffer
+    /// Returns the number of entries missed before those that were read
+    pub fn read(&mut self, rbuf: &mut Vec<WholeEntry<E>>) -> Result<u64, S::Error> {
+        let pre_overwrite_seqn = self.snap_overwrite_seqn()?;
+        let pre_write_seqn = self.snap_write_seqn()?;
+
+        // If writer has overwritten unread entries between reads, store missed and correct read_seqn
+        let mut n_missed_before_read = num_missed(self.read_seqn, pre_overwrite_seqn);
+        self.read_seqn += n_missed_before_read;
+        // If any entries were missed and there is a stored prefix, then the entry after the prefix was missed.
+        // The prefix is dropped and added to the missed count.
+        // Note: the read sequence number is updated above because the prefix was included in the last read
+        if Into::<u64>::into(n_missed_before_read) > 0 && self.drop_prefix() {
+            n_missed_before_read += 1;
+        }
+
+        // Perform reads into snapshot buffer up to write sequence number
+        let first_read_seqn = self.read_seqn;
+        let mut buf_snapshot: Vec<E> = Vec::new();
```

Comment:
To avoid allocating on what might be a relative hot read path and you don't want to avoid an intermediate buffer entirely in order to get bulk reads, I would recommend instead storing a `buf_snapshot` field on the `RaceReader` struct and just `clear`ing it here instead of new'ing up a new one.